### PR TITLE
MonotonicSplineTransformer

### PR DESCRIPTION
**Pull request by koaning**
An implementation for https://github.com/koaning/scikit-lego/issues/704

### CODE REVIEW

**Comment by koaning** [comment]
@matekadlicsko will present this to the larger crowd tomorrow.

**Comment by matekadlicsko** [comment]
I believe this implementation has an error. The cumsum should go over the columns instead of the rows, which in turn makes sorting obsolete as well. I can fix this.

**Comment by koaning** [comment]
Feel free to put a PR on top of this one. In particular, if you could contribute a unit test that demonstrates when something breaks that would be stellar.

**Comment by koaning** [comment]
@FBruzzesi I think all the base tests pass, so this one might be green to go in.

**Review by FBruzzesi — COMMENTED** [review]
Hey @koaning, I left a bunch of comments, mostly for documenting and some other minor details.

For the algorithm itself I did not do my homework properly as I did not have the chance to watch last Friday live stream, yet I trust your and Mate Kadlicsko implementation and review.

File: docs/_scripts/preprocessing.py [hunk 1]
```diff
@@ -325,3 +325,67 @@ def generate_dataset(start, n=600):
...
+
+plt.figure(figsize=(10, 4))
+plt.scatter(X, y)
+# --8<-- [end:monotonic-spline]
+
+plt.savefig(_static_path / "monotonic-spline.png")
+plt.clf()
+
+# --8<-- [start:monotonic-spline-transform]
+from sklearn.preprocessing import SplineTransformer
```
Comment (koaning) [inline docs/_scripts/preprocessing.py:350]:
```suggestion
```

File: tests/test_preprocessing/test_monospline.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+@pytest.mark.parametrize("knots", ["uniform", "quantile"])
+def test_monotonic_spline_transformer(n_knots, degree, knots):
+    X = np.random.uniform(size=(100, 10))
+    transformer = MonotonicSplineTransformer(n_knots=n_knots, degree=degree, knots=knots)
+    transformer_sk = SplineTransformer(n_knots=n_knots, degree=degree, knots=knots)
+    transformer.fit(X)
+    transformer_sk.fit(X)
+    out = transformer.transform(X)
+    out_sk = transformer_sk.transform(X)
+    print(out.shape, out_sk.shape)
```
[the remark this answers is not in the document]
Comment (koaning) [reply tests/test_preprocessing/test_monospline.py:25]: Good question. Lemme explore that.

File: docs/api/preprocessing.md [hunk 3]
```diff
@@ -69,3 +69,8 @@
     options:
         show_root_full_path: true
         show_root_heading: true
+
+:::sklego.preprocessing.monotonicspline.MonotonicSplineTransformer
+    options:
+        show_root_full_path: true
+        show_root_heading: true
```
[the remark this answers is not in the document]
Comment (koaning) [reply docs/api/preprocessing.md:76]: Fair! Will make the change.

File: sklego/preprocessing/monotonicspline.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            estimator=self,
+        )
+        if X.shape[1] != self.n_features_in_:
+            raise ValueError("Number of features going into .transform() do not match number going into .fit().")
+
+        out = []
+        for col in range(X.shape[1]):
+            out.append(
+                np.cumsum(
+                    self.spline_transformer_[col].transform(X[:, [col]])[:, ::-1],
```
[the remark this answers is not in the document]
Comment (matekadlicsko) [reply sklego/preprocessing/monotonicspline.py:96]: The direction in which you perform the cumsum on the columns will determine whether the output splines will be increasing or decreasing. However, now that you mention it, I'm pretty sure that
```
1 - np.cumsum(without inversion)
```
would achieve the same effect.

File: tests/test_preprocessing/test_monospline.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (koaning) [reply tests/test_preprocessing/test_monospline.py]: We hadn't configured T201 it seems. Fixed now.

File: sklego/preprocessing/monotonicspline.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    """The `MonotonicSplineTransformer` integrates the output of the `SplineTransformer` in an attempt to make monotonic features.
+
+    This estimator is heavily inspired by [this blogpost](https://matekadlicsko.github.io/posts/monotonic-splines/) by Mate Kadlicsko.
+
+    Parameters
+    ----------
+    n_knots : int, default=3
+        The number of knots to use in the spline transformation.
+    degree : int, default=3
+    knots: str, default="uniform"
```
[the remark this answers is not in the document]
Comment (FBruzzesi) [reply sklego/preprocessing/monotonicspline.py:18]: I think this is still relevant, aside that it seems ready to merge

**Review by FBruzzesi — APPROVED** [review]
🙌🏼

**Pull request merged**